<h1 style="text-align: center; color: red; font-size: 25pt; font-weight: bold; text-shadow: 1px 1px 1px rgb(50, 100, 100)"> Terror Attacks In Nigeria .... </h1> 
<p style="font-size: 12pt; text-align: center"> Abubakar Abdulkadir</p>

<img src="images/bg.jpg" style="width: 100%; height:350px; display: inline-block" />

<h1 style='background-color: red; padding: 10px; color: white'> 1.0 Project Overview </h1>

```A Terror attack is the unlawful use of force or violence against persons or property to intimidate or coerce a government, the civilian population, or any segment thereof in furtherance of political or social objectives"``` - - The Federal Bereau of Intelligence, US.

Nigeria is one of such countries of the world where insecurity and terrorist activities has been on the rise. The West African nation is one of the most hazardous nations in the world, with the most frequent and deadly attacks. There have been 2470 terrorist incidents in the last five years, resulting in 17751 fatalities and 7345 injuries (WorldData.info). With the emergence of new dangerous groups like the Fulani extremists and herdsmen, these numbers continue to rise daily.

Attacks in Nigeria have the notable characteristic that each terrorist group or sect's attacks can almost immediately be linked to them due to similarities in the attack's mode, target, duration, motive, and casualty toll, among other factors. It can be argued that patterns and intelligence can be extracted from datasets of these attacks. Even if these records are accessible in public places like Wikipedia, social media platforms like Facebook and Twitter, news websites, and newspapers, to name a few, they aren't offered in formats that are simple to use with tools for data manipulation and analysis. Consequently, limiting the utilizing of data-centric approach in combating terrorism in the nation.

This furthers the project's objectives to aid in the fight against insecurity in Nigeria by 
- collecting data about attacks and security incidences in Nigeria
- processing the data into easy to use formats
- Performing analysis on the dataset to identify trends and patterns
- providing answers to questions about the mode of operation, timing, motive, and other aspects of the attacks.
- Building Smart Security Systems for predicting insecurity incidences in Nigeria and making recommendations to prevent future occurences 

Although the project's main focus is on insecurity in Nigeria, it has worldwide significance because terrorism has always been a problem on a global scale. While it is true that this threat's exposure differs from region to region and country to country, the necessity of taking effective action on a global scale cannot be overstated.

<h1 style='background-color: red; padding: 10px; color: white'> 2.0 Data Collection </h1>

The datasets considered for this project are collected from four different sources including the Armed Conflict Location and Event Data project (ACLED), the global terrorism database (GTD), CHATGPT API and GITHUB 

### 2.1 importing the datasets

In [22]:
df_acled = pd.read_csv('datasets/uncleaned/acled-Nigeria 2009.csv') # ACLED dataset 
df_gtd_2020 = pd.read_excel('datasets/uncleaned/gtd-2020.xlsx') # GTD dataset 1970 to 2020
df_gtd_2021 = pd.read_excel('datasets/uncleaned/gtd-jan 2021.xlsx') # GTD dataset Jan 2021 to Jun 2021
df_lga = pd.read_csv('datasets/uncleaned/lgas.csv') # LGA datasets from github

### 2.2 Armed Conflict Location & Event Data Project (ACLED)

The dataset from this source includes All insecurity incidences in Nigeria from January 2009 to March 28 2023. It has 31 features which includes time and date related features, location features, event properties, casualties, information sources, and brief description of event. The dataset comes in a CSV format and can be assessed on the ACLED data download <a href='https://acleddata.com/data-export-tool/'>portal</a>

In [50]:
df_acled.head()

event_id_cnty     event_date  year  time_precision       disorder_type  \
0      NIG30571  24 March 2023  2023               1      Demonstrations   
1      NIG30577  24 March 2023  2023               1      Demonstrations   
2      NIG30583  24 March 2023  2023               1      Demonstrations   
3      NIG30608  24 March 2023  2023               1      Demonstrations   
4      NIG30559  23 March 2023  2023               1  Political violence   

                   event_type             sub_event_type  \
0                    Protests           Peaceful protest   
1                    Protests  Protest with intervention   
2                    Protests           Peaceful protest   
3                    Protests           Peaceful protest   
4  Violence against civilians                     Attack   

                               actor1                   assoc_actor_1  inter1  \
0                Protesters (Nigeria)                             NaN       6   
1                Protesters (Nigeria)                             NaN       6   
2                Protesters (Nigeria)                 Women (Nigeria)       6   
3                Protesters (Nigeria)  PDP: People's Democratic Party       6   
4  Unidentified Armed Group (Nigeria)                             NaN       3   

   ...      location latitude  longitude  geo_precision  \
0  ...  Ayede-Ogbese   7.2057     5.3875              1   
1  ...           Owo   7.1962     5.5868              2   
2  ...       Akwanga   8.9167     8.3833              1   
3  ...          Karu   9.0469     7.7636              1   
4  ...         Gboko   7.3179     8.9951              1   

                                 source       source_scale  \
0                    Vanguard (Nigeria)           National   
1                         Nigeria Punch           National   
2  Sahara Reporters; Vanguard (Nigeria)  National-Regional   
3     Sun (Nigeria); Vanguard (Nigeria)           National   
4          Nigeria Punch; Sun (Nigeria)           National   

                                               notes fatalities  \
0  On 24 March 2023, youths and residents of Ayed...          0   
1  On 24 March 2023, youths of Uso community prot...          0   
2  On 24 March 2023, women in Akwanga (Akwanga, N...          0   
3  On 24 March 2023, a large crowd of supporters ...          0   
4  On 23 March 2023, an unidentified armed group ...          0   

                       tags   timestamp  
0      crowd size=no report  1679938033  
1      crowd size=no report  1679938033  
2      crowd size=no report  1679938034  
3  crowd size=a large crowd  1679938034  
4                       NaN  1679938033  

[5 rows x 31 columns]

In [57]:
df_acled.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28160 entries, 0 to 28159
Data columns (total 31 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   event_id_cnty       28160 non-null  object 
 1   event_date          28160 non-null  object 
 2   year                28160 non-null  int64  
 3   time_precision      28160 non-null  int64  
 4   disorder_type       28160 non-null  object 
 5   event_type          28160 non-null  object 
 6   sub_event_type      28160 non-null  object 
 7   actor1              28160 non-null  object 
 8   assoc_actor_1       8139 non-null   object 
 9   inter1              28160 non-null  int64  
 10  actor2              21306 non-null  object 
 11  assoc_actor_2       6528 non-null   object 
 12  inter2              28160 non-null  int64  
 13  interaction         28160 non-null  int64  
 14  civilian_targeting  10929 non-null  object 
 15  iso                 28160 non-null  int64  
 16  regi

The dataset contains 28160 insecurity incidences cutting accross different events such as Violence against civilians, Battles, Protests, Riots, Explosions/Remote violence and Strategic developments. It has 3 floatinf point features, 9 integers and 19 objects.  

In [59]:
#save dateset into the cleaned folder
df_acled.to_csv('datasets/cleaned/df_acled_nig.csv', index=False)

### 2.2 Global Terrorism Database (GTD)

The Global Terrorism Database (GTD) is an open-source database including information on terrorist events around the world from 1970 through 2020 (with annual updates planned for the future). Unlike many other event databases, the GTD includes systematic data on domestic as well as international terrorist incidents that have occurred during this time period and now includes more than 200,000 cases. Although, for this use case, the data collected from this database was filtered to contain only events which occured in Nigeria. The dataset can be downloaded from their website at <a href="https://www.start.umd.edu/gtd/"> GTD portal</a> 

#### 2.2.1 Dataset 2020

In [24]:
df_gtd_2020.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 209706 entries, 0 to 209705
Columns: 135 entries, eventid to related
dtypes: datetime64[ns](1), float64(54), int64(23), object(57)
memory usage: 216.0+ MB


In [25]:
df_gtd_2020.shape

(209706, 135)

In [26]:
df_gtd_2020.head()

eventid  iyear  imonth  iday approxdate  extended resolution  country  \
0  197000000001   1970       7     2        NaN         0        NaT       58   
1  197000000002   1970       0     0        NaN         0        NaT      130   
2  197001000001   1970       1     0        NaN         0        NaT      160   
3  197001000002   1970       1     0        NaN         0        NaT       78   
4  197001000003   1970       1     0        NaN         0        NaT      101   

          country_txt  region  ... addnotes scite1 scite2  scite3  dbsource  \
0  Dominican Republic       2  ...      NaN    NaN    NaN     NaN      PGIS   
1              Mexico       1  ...      NaN    NaN    NaN     NaN      PGIS   
2         Philippines       5  ...      NaN    NaN    NaN     NaN      PGIS   
3              Greece       8  ...      NaN    NaN    NaN     NaN      PGIS   
4               Japan       4  ...      NaN    NaN    NaN     NaN      PGIS   

   INT_LOG  INT_IDEO INT_MISC INT_ANY  related  
0        0         0        0       0      NaN  
1        0         1        1       1      NaN  
2       -9        -9        1       1      NaN  
3       -9        -9        1       1      NaN  
4       -9        -9        1       1      NaN  

[5 rows x 135 columns]

In [28]:
df_gtd_2020.iyear.unique()

array([1970, 1971, 1972, 1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980,
       1981, 1986, 1982, 1983, 1984, 1985, 1987, 1988, 1989, 1990, 1991,
       1992, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003,
       2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014,
       2015, 2016, 2017, 2018, 2019, 2020], dtype=int64)

This dataset contains 135 features comprising of 1 datetime feature, 54 floating point features, 23 integer features and 57 object features. It contains 209,706 events from 1970 to 2020. Although, this are all events which occured accross the globe.
Hence, there is a need to filter the dataset to contain only incidences which occured in Nigeria

In [39]:
# filtering the dataset to contain only events which occured in Nigeria
df_gtd_nig_2020 = df_gtd_2020[df_gtd_2020['country_txt'] == 'Nigeria']
df_gtd_nig_2020.shape

(5550, 135)

There are 5550 insecurity incidences recorded in Nigeria within this time frame.

#### 2.2.2 Dataset 2021

This is similar to the GTD 2020 dataset. The difference is that this contains events occuring between jan 2021 to jun 2021.

In [42]:
# filtering the dataset to contain only events which occured in Nigeria
df_gtd_nig_2021 = df_gtd_2021[df_gtd_2021['country_txt'] == 'Nigeria']
df_gtd_nig_2021.shape

(448, 135)

Within this time frame, there were 448 recorded incidences

#### 2.2.2 Merging both Datasets

In [44]:
df_gtd_nig = pd.concat([df_gtd_nig_2020, df_gtd_nig_2021])
df_gtd_nig.shape

(5998, 135)

Therefore, the GTD dataset recorded from 1970 to 2021, a total of 5998 security incidences in Nigeria.  

In [48]:
# save dataset as a new csv file
df_gtd_nig.to_csv('datasets/cleaned/df_gtd_nig.csv', index=False)

### 2.3 ChatGPT API

ChatGPT API is the Application interface through which the chatGPT model can be accessed. ChatGPT is an AI chatbot that was initially built on a family of large language models (LLMs) collectively known as GPT-3. This model has been trained on huge amount of data to understand and generate human like responses to text prompts. Although, it occassional provides incorrect responses, its ability to produce human-like, and frequently accurate, responses to a vast range of questions is why it is considered for this pproject.

This API is used to curate a list of properties for each local government area in Nigeria. Alhough, it is worthy of note that whatever response the API is providing is based of its knowledge of 2021 which is the range of dataset it has been trained on.

#### 2.3.1 Collecting Dataset from Chat GPT API

In [5]:
#importing packages
import openai
import pandas as pd
from csv import writer
import re

In [2]:
openai.api_key = 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'  # input the API key

In [4]:
df = pd.read_csv('datasets/cleaned/lgas.csv') # read csv file of all local government area in Nigeria

In [ ]:
# seek details in batches of 100
for i in range(0, 800, 100):
    cur_lgas = df.iloc[i:i + 100]
    lga = list(cur_lgas['name'])
    
    #loop through each local government name in each batch
    for j in range(len(lga)):
        prompt = '''In one word each, Provide actual values for the questions  
                1. Percentage of Educated
                2. Level of Isolation
                3. Natural Barriers
                4. Elevation
                5. Population
                6. Major Ocupation
                7. Political stability
                8. Economic Situation
                9. Average Cost of living
                10. Employment rate
                11. Ethnic Marginalization
                12. Transport infrastructure
                13. Longitude 
                14. Latitude
                15. Income Equality
                16. Major ethnic group
                17. Communication network
                18. Infrastructure development
                19. Dominant Age group
                20. Average family size
                21. Average education level
                22. Average Income in Naira
                of ${}. 
                Numper your reponse and put your answer in a single line '''.format(lga[j])

        response = openai.ChatCompletion.create(
              model="gpt-3.5-turbo",
              messages=[ 
                    {"role": "user", "content": prompt},     
                ]
        )
        
        data = (response['choices'][0]['message'].content).replace('\n', ' ')

        # append the response to a file
        f_object = open('laga_dataset.txt', 'a', encoding="utf-8")
        f_object.write("*** 0" + lga[j] + " "  + " " + data + " " + "*** /n")

#### 2.3.2 Formating the dataset into CSV

In [67]:
#read the txt file
with open('datasets/uncleaned/laga_dataset.txt', 'r') as f_handler:
    features = f_handler.read()

# split the dataset into each local goverment with the "*** 0" marker and ignore the first entry (It is empty)
each_feature = features.split("*** 0")[1:]

# split each feature into a list and append it to another list to form a list of list
features_list = []
for feature in each_feature:
    feature = feature.replace("*** /n", "")
    str_list = re.sub(r'\d+\. ', '<break>', feature).split('<break>')
    str_list = [x.strip() for x in str_list if (len(x.strip()) > 1 or re.findall('\d+', x))]
    features_list.append(str_list)

In [68]:
# define the pandas column header names
columns = [
            "Lga Name", "percentage of educated", "Level of Isolation", "Natural Barriers", "Elevation", 
            "Population", "Major Ocupation", "Political stability", 
            "Economic Situation", "Average Cost of living", "Employment rate", 
            "Ethnic Marginalization", "Transport infrastructure", "Longitude",  
            "Latitude", "Income Equality", "Major ethnic group", "Communication network", 
            "Infrastructure development", "Dominant Age group", "Average family size",
            "Average education level", "Average Income in Naira"
]


#create a dataframe with the dataset
feature_df = pd.DataFrame(features_list, columns=columns)

In [69]:
feature_df.head()

Lga Name percentage of educated Level of Isolation  Natural Barriers  \
0  Aba North                    70%                Low              None   
1  Aba South                    70%                Low              None   
2  Arochukwu                    70%                Low  Hills and Rivers   
3      Bende                    90%                Low            Rivers   
4    Ikwuano                    70%                Low  Hills and Rivers   

      Elevation Population       Major Ocupation Political stability  \
0  104m (341ft)    382,784  Trading and Commerce   Relatively Stable   
1           30m  1 million               Trading   Relatively stable   
2        1500ft    150,000           Agriculture   Relatively stable   
3    200 meters    100,000               Farming              Stable   
4          190m    150,000               Farming   Relatively Stable   

  Economic Situation       Average Cost of living  ... Longitude  Latitude  \
0         Struggling  ₦120,000 ($312) per month  ...  7.39° E  5.05° N   
1         Struggling                    ₦50,000  ...    7.37 E    5.53 N   
2         Developing                    ₦40,000  ...     7.9 E     5.5 N   
3         Struggling                    ₦50,000  ...   7.5° E   5.6° N   
4         Struggling             ₦50,000 ($125)  ...    7.5°E    5.4°N   

  Income Equality Major ethnic group Communication network  \
0             Low               Igbo                  Good   
1             Low               Igbo      Well-established   
2             Low               Igbo               Limited   
3             Low               Igbo      Well established   
4             Low                Ibo               Limited   

  Infrastructure development Dominant Age group Average family size  \
0                 Developing        25-44 years            4 people   
1                    Limited              25-44                   4   
2                       Slow              Youth                   4   
3       Moderately developed        25-40 years                   5   
4                 Developing              Youth                   4   

  Average education level      Average Income in Naira  
0     High school diploma  ₦150,000 ($390) per month  
1             High school                    ₦40,000  
2     High school diploma                    ₦50,000  
3        Secondary school                    ₦80,000  
4             High School             ₦60,000 ($150)  

[5 rows x 23 columns]

In [70]:
feature_df.shape

(731, 23)

In [71]:
feature_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 23 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Lga Name                    731 non-null    object
 1   percentage of educated      730 non-null    object
 2   Level of Isolation          730 non-null    object
 3   Natural Barriers            730 non-null    object
 4   Elevation                   730 non-null    object
 5   Population                  730 non-null    object
 6   Major Ocupation             730 non-null    object
 7   Political stability         730 non-null    object
 8   Economic Situation          730 non-null    object
 9   Average Cost of living      730 non-null    object
 10  Employment rate             730 non-null    object
 11  Ethnic Marginalization      730 non-null    object
 12  Transport infrastructure    730 non-null    object
 13  Longitude                   730 non-null    object

Although, there are 774 local government areas in Nigeria. We could only extract records for 731. There is no feature record for 43 local governments. All the features are returned as object types.

In [72]:
# save this dataset into the cleaned folder
feature_df.to_csv('datasets/cleaned/df_lga_features.csv', index=False)

### 2.4 GitHub

The local governments in Nigeria dataset was gotten from github. It contains all the 774 local government area in Nigeria. The Local goverment is the lowest level of governemnt in Nigeria. Hence, the reason why properties of places are collected in the group of local governments. The dataset was gotten from <a href='https://github.com/xosasx/nigerian-local-government-areas'> this </a> repository. 

In [61]:
df_lga.head()

id       name  state_id state_code state_name  country_id country_code  \
0   1  Aba North       303         AB       Abia         161           NG   
1   2  Aba South       303         AB       Abia         161           NG   
2   3  Arochukwu       303         AB       Abia         161           NG   
3   4      Bende       303         AB       Abia         161           NG   
4   5    Ikwuano       303         AB       Abia         161           NG   

  country_name  latitude  longitude wikiDataId  
0      Nigeria  5.333333   7.316667   Q4663114  
1      Nigeria  5.100000   7.350000   Q4663117  
2      Nigeria  5.383333   7.916667    Q965036  
3      Nigeria  5.566667   7.633333   Q4887027  
4      Nigeria  5.433333   7.566667   Q5996447

In [62]:
df_lga.shape

(774, 11)

In [64]:
df_lga.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 774 entries, 0 to 773
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id            774 non-null    int64  
 1   name          774 non-null    object 
 2   state_id      774 non-null    int64  
 3   state_code    761 non-null    object 
 4   state_name    774 non-null    object 
 5   country_id    774 non-null    int64  
 6   country_code  774 non-null    object 
 7   country_name  774 non-null    object 
 8   latitude      774 non-null    float64
 9   longitude     774 non-null    float64
 10  wikiDataId    774 non-null    object 
dtypes: float64(2), int64(3), object(6)
memory usage: 66.6+ KB


In [65]:
# save this dataset into the cleaned folder
df_lga.to_csv('datasets/cleaned/df_lga_nig.csv', index=False)